### ISSUE TITLE
match clause does not handle LDAP attribute lookups returning multivalues.

### ISSUE THREAD

**Issue by Mutisk-zz** [rabbitmq/rabbitmq-auth-backend-ldap#100]
* RabbitMQ version 3.7.8
* Erlang version 21.1
sample resource access query:

{match, {string, "RMQ-${vhost}-${resource}-${name}-${permission}"},{attribute, "${user_dn}" , "postOfficeBox" } }

When this is single-valued in ldap DB, results are as expected. 
`LDAP evaluated match "RMQ-websocket-exchange-pluupp-configure" against RE "RMQ-websocket-.*-.*-.*": true`

However when adding another postOfficeBox-attribute the evaluation changes to:
`LDAP evaluated match "RMQ-x-.*-.*-.*; RMQ-websocket-.*-.*-.*" against RE "RMQ-websocket-exchange-pluupp-configure": false`

Not only is the result turned false, but also the order of the comparison is reversed. My regular expressions are treated as it were the input string.  The culprit of this is do_match_bidirectionally, and I have a pull request for a patch that I think address this.

(PS. I also think I found some dead code in a bunch of evaluate0-function for match. They all result in the same evaluate call despite having different guards. I'm thinking the guards would make more sense to the results being lists, rather than the queries? Where is a good place to discuss this? )

**Comment by michaelklishin** [comment]
The mailing list is a great place to discuss anything you have in mind. Thank you for looking into a PR, please update the integration test suite (which currently uses an OpenLDAP node in a Vagrant VM) because I am not sure how matching for multi-value responses should work exactly (there's more than one to do it).

**Comment by michaelklishin** [comment]
The PR is in #101.

**Comment by lukebakken** [comment]
@Mutisk - if you can, please provide the log output when the `network_unsafe` setting is in use, or provide a way to populate an LDAP instance with data that can reproduce this issue (an ldif file with data would be great).

**Comment by Mutisk-zz** [comment]
@michaelklishin 
Yeah, true. I also removed the bidirectional function but did not look into tests(nor remove it there). I haven't used vagrant before, but I will give it a try when I get the chance. Regarding how multivalues should work, I wanted authorization to succeed if any of the values in the ldap db would match my access query. 

But if you say there are more than one way to do it, what kind of use case would not want "true if at least one of values satisfy"?

The current implementation is buggy btw since bidirectional match is not okay for regular expressions, and any match would rely on the order they get concatenated. On the other hand, I'm not sure what the rationale behind bidirectional match is even. It is tried after a normal match fails, if the length is sufficient - but why?. It looks incorrect and exploitable, and the intent is not that well documented afaict

**Comment by Mutisk-zz** [comment]
I got the vagrant box up and could run the test suite unmodified (the do_match_bidirectional test still passes despite me renaming that function to do_match_multi). But writing a ldap integration test for my code is beyond me at this point.

To reproduce is pretty simple: configure your rabbitmq to do ldap resource_access_query in advanced.config
```
[
{rabbitmq_auth_backend_ldap,
        [
        {tag_queries,
                [
                {administrator, {constant, true}}
                ]
        },
        {resource_access_query,
                            {match, {string, "RMQ-${vhost}"},{attribute, "${user_dn}" , "postOfficeBox" } }
        }
        ]
}
].

```

Then add to your user_dn an attribute postOfficeBox with a regex value: ^RMQ-.*$
Results: All access to create/read/write on resources are granted.
Add a second value of postOfficeBox (thus making it multi-value): ^RMQ-test.*$
Results: All access to resources are denied, ldap network log show it reverse RE and String in the match: "^RMQ-.*$;^RMQ-test.*$" versus RE "RMQ-${vhost}"

In my case when $vhost=websocket:
`[info] <0.317.0>     LDAP evaluated match "^RMQ-test.*$; ^RMQ-.*$" against RE "RMQ-websocket": false`

Code analysis at line 373 (do_match_bidirectionally) shows it previously tries and the arguments in the correct order but, fails silently with no logging, then does the RE<-->String mixup and fails again which is logged. 

My PR handles multivariable attribute lookups by "pass if any match, fail if none match" which should be the most common use case. The only other way that makes sense would be to require "pass if all match, or fail" but then you are building a blacklist.  It will probably also mean you can clean out dead code, since do_match_multi behaves well even when you have regular string arguments.

**Comment by lukebakken** [comment]
I'll see what I can do about getting a breaking test into the suite that is then fixed by this.

**Comment by lukebakken** [comment]
I've added a test that demonstrates the issue in this branch:

https://github.com/rabbitmq/rabbitmq-auth-backend-ldap/tree/rabbitmq-auth-backend-ldap-100-test

I'll merge this into #101.

**Comment by lukebakken** [comment]
Fixed by #101
